core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -90,30 +93,239 @@
+  private long mCurrentBlockEndPos = UNINITIALIZED_BLOCK_POS;
   /** Current block in stream backing this stream. */
   protected BlockInStream mCurrentBlockInStream;
   /**
    * Current block out stream writing the data into the local worker. This is only used when the in
    * stream reads from a remote worker.
    */
   protected BlockOutStream mCurrentCacheStream;
-  /** The blockId used in the block streams. */
-  private long mStreamBlockId;
 
   /** The read buffer in file seek. This is used in {@link #readCurrentBlockToEnd()}. */
   private final byte[] mSeekBuffer;
 
   /**
+   * States related to position that need to be in sync.
+   */
+  protected class PositionState {
+    /** Block ID used when position reaches EOF, should be < 0. */
+    public static final long EOF_BLOCK_ID = -1;
+
+    /** File length. */
+    private final long mFileLength;
+    /** Block IDs. */
+    private final List<Long> mBlockIds;
+    /** Block size in bytes. */
+    private final long mBlockSize;
+
+    /**
+     * Current position of the file.
+     * Initially, it is 0.
+     * When EOF is reached, it is the file length.
+     */
+    private long mPos;
+    /** The remaining bytes of the file. */
+    private long mRemaining;
+
+    // States derived from mPos.
+    /**
+     * Index of the block where the current file position belongs to.
+     * Initially, it is 0.
+     * When EOF is reached, it is the size of mStatus.getBlockIds.
+     */
+    private int mBlockIndex;
+    /**
+     * ID of the block where the current file position belongs to.
+     * Initially, it is the first block ID.
+     * When EOF is reached, it is {@link #EOF_BLOCK_ID}.
+     */
+    private long mBlockId;
+    /**
+     * Start position of the block where the current file position belongs to.
+     * Initially, it is 0.
+     * When EOF is reached, it is the file length.
+     */
+    private long mBlockStartPos;
+    /**
+     * Start position of the next block.
+     * Initially, it is the start position of the second block, or the file length if there is only
+     * one block.
+     * When EOF is reached or the current block is the last block, it is the file length.
+     */
+    private long mNextBlockStartPos;
+
+    /**
+     * Creates a new position state.
+     *
+     * @param fileLength the file length in bytes
+     * @param blockIds the block IDs, sorted by the blocks to be read in as a stream
+     * @param blockSize the block size in bytes
+     */
+    public PositionState(long fileLength, List<Long> blockIds, long blockSize) {
+      mFileLength = fileLength;
+      mBlockIds = blockIds;
+      mBlockSize = blockSize;
+
+      mPos = 0;
+      mRemaining = mFileLength;
+      computeState();
+    }
+
+    /**
+     * Adds position by increment, updates the block index and block position if necessary.
+     *
+     * @param increment the position increment, expected to be non-negative
+     */
+    public void addPos(long increment) {
```

Comment:
nit: how about naming this `incrementPos(long increment)`

Reply:
renamed to `increase`